common/common.go

```diff
@@ -433,28 +433,78 @@ func InArray(val, array interface{}) (exists bool, index int) {
 	return
 }
 
-// Errors defines multiple errors
-type Errors []error
+// AppendError appends error in a more idiomatic way. This can start out as a
+// standard error e.g. err := errors.New("random error")
+// err = AppendError(err, errors.New("another random error"))
+func AppendError(original, incoming error) error {
+	errSliceP, ok := original.(*multiError)
+	if ok {
+		errSliceP.offset = 0
+	}
+	if incoming == nil {
+		return original // Skip append - continue as normal.
+	}
+	if !ok {
+		// This assumes that a standard error is passed in and we can want to
+		// track it and add additional errors.
+		errSliceP = &multiError{}
+		if original != nil {
+			errSliceP.loadedErrors = append(errSliceP.loadedErrors, original)
+		}
+	}
+	if incomingSlice, ok := incoming.(*multiError); ok {
+		// Join slices if needed.
+		errSliceP.loadedErrors = append(errSliceP.loadedErrors, incomingSlice.loadedErrors...)
+	} else {
+		errSliceP.loadedErrors = append(errSliceP.loadedErrors, incoming)
+	}
+	return errSliceP
+}
 
-// Error implements error interface
-func (e Errors) Error() string {
-	if len(e) == 0 {
-		return ""
+// multiError holds all the errors as a slice, this is unexported so it forces
+// ibuilt error handling.
+type multiError struct {
+	loadedErrors []error
+	offset       int
+}
+
+// Error displays all errors comma separated, if unwrapped has been called and
+// has not been reset will display the individual error
+func (e *multiError) Error() string {
+	if e.offset > 0 {
+		last := e.offset - 1
+		if last >= len(e.loadedErrors) {
+			return ""
+		}
+		return e.loadedErrors[last].Error()
 	}
-	var r string
-	for i := range e {
-		r += e[i].Error() + ", "
+	allErrors := make([]string, len(e.loadedErrors))
+	for x := range e.loadedErrors {
+		allErrors[x] = e.loadedErrors[x].Error()
 	}
-	return r[:len(r)-2]
+	return strings.Join(allErrors, ", ")
 }
 
-// Unwrap implements interface behaviour for errors.Is() matching NOTE: only
-// returns first element.
-func (e Errors) Unwrap() error {
-	if len(e) == 0 {
-		return nil
+// Unwrap increments the offset so errors.Is() can be called to its individual
+// error for correct matching.
+func (e *multiError) Unwrap() error {
```

Comment:
I'm getting a little confused on your implementation. I'm unsure how its working within Go
Check out the following, this does not touch your code:
```go
func TestUnwrap34(t *testing.T) {
	errOne := errors.New("one")
	errTwo := errors.New("two: ")
	errThree := errors.New("three: ")

	errTwoOnErrOne := fmt.Errorf("%v %w", errTwo, errOne)
	errThreeOnErrTwoOnErrOne := fmt.Errorf("%v %w", errThree, errTwoOnErrOne)
	errTwoUnwrappedFromThree := errors.Unwrap(errThreeOnErrTwoOnErrOne)
	t.Log(errTwoUnwrappedFromThree)
	errOneUnwrappedFromTwo := errors.Unwrap(errTwoUnwrappedFromThree)
	t.Log(errOneUnwrappedFromTwo)
	if errors.Is(errOneUnwrappedFromTwo, errOne) {
		t.Log("it sure is")
	}
}

```
However, using your way
```go
func TestUnwrap600(t *testing.T) {
	errOne := errors.New("one")
	errTwo := errors.New("two: ")
	errThree := errors.New("three: ")

	errTwoOnErrOne := AppendError(errOne, errTwo)
	errThreeOnErrTwoOnErrOne := AppendError(errTwoOnErrOne, errThree)
	errTwoUnwrappedFromThree := errors.Unwrap(errThreeOnErrTwoOnErrOne)
	if errors.Is(errTwoUnwrappedFromThree, errThree) {
		t.Logf("should I be error three?")
	}
}

```

Reply:
Yeah I didn't see it necessary to remove the error when unwrapping, I guess I can return a minimized struct when calling unwrap if you want?

Reply:
That kind of logic is fine, we're doing a multi error after all. But then, why isn't the error also an error one?
```
func TestUnwrap600(t *testing.T) {
	errOne := errors.New("one")
	errTwo := errors.New("two: ")
	errThree := errors.New("three: ")

	errTwoOnErrOne := AppendError(errOne, errTwo)
	errThreeOnErrTwoOnErrOne := AppendError(errTwoOnErrOne, errThree)
	errTwoUnwrappedFromThree := errors.Unwrap(errThreeOnErrTwoOnErrOne)
	if !errors.Is(errTwoUnwrappedFromThree, errOne) {
		t.Logf("why aren't I err one?")
	}
}
```

Reply:
The way it unwraps it increments until it hits the last one then resets. So its designed to be unwrapped in errors package when calling errors.Is()

Reply:
I think I got a bit lost. Don't worry about this one, however the other comment may be a little bug to fix. Then I think its all good